src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,300 @@
+                }
+            }
+
+            @Override
+            public void onCompleted() {
+
+            }
+        };
+
+        /**
+         * Start heartbeat schedule to keep alive leases
+         */
+        keepAliveSchedule = Executors.newSingleThreadScheduledExecutor();
+        keepAliveSchedule.scheduleAtFixedRate(
+                ()-> keepAliveExecutor()
+                , 0, 500, TimeUnit.MILLISECONDS);
+        /**
+         * Start heartbeat schedule to remove dead leases
+         */
+        deadLineSchedule = Executors.newSingleThreadScheduledExecutor();
+        deadLineSchedule.scheduleAtFixedRate(
+                ()-> deadLineExecutor()
+                , 1000, 1000, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * New a lease with ttl value
+     * @param ttl ttl value
+     * @return
+     */
+    @Override
+    public ListenableFuture<LeaseGrantResponse> grant(long ttl) {
+        LeaseGrantRequest leaseGrantRequest =
+                LeaseGrantRequest.newBuilder()
+                        .setTTL(ttl)
+                        .build();
+        return leaseFutureStub.leaseGrant(leaseGrantRequest);
+
+    }
+
+    /**
+     * revoke one lease and the key bind to this lease will be removed
+     * @param leaseId the id of lease
+     * @return
+     */
+    @Override
+    public ListenableFuture<LeaseRevokeResponse> revoke(long leaseId) {
+        LeaseRevokeRequest leaseRevokeRequest =
+                LeaseRevokeRequest.newBuilder()
+                        .setID(leaseId)
+                        .build();
+        return leaseFutureStub.leaseRevoke(leaseRevokeRequest);
+    }
+
+    /**
+     * keep alive one lease in background
+     * @param leaseId
+     */
+    @Override
+    public void keepAlive(long leaseId) {
+        synchronized (lock){
+            if(keepAlives.containsKey(leaseId)){
+
+            }else{
+                Lease lease = new Lease(leaseId);
+                lease.setNextKeepAlive(System.currentTimeMillis());
+                lease.setDeadLine(System.currentTimeMillis() + firstKeepAliveTimeOut);
+                keepAlives.put(leaseId, lease);
+            }
+        }
+
+        return;
+    }
+
+    /**
+     * keep alive one lease in background
+     * @param leaseId
+     */
+    @Override
+    public ListenableFuture<LeaseKeepAliveResponse> keepAliveOnce(long leaseId) {
+
+        /**
+         * to be completed, I will return a ListenableFuture value in the future
+         */
+        StreamObserver<LeaseKeepAliveRequest> requestObserver =
+                leaseStub.leaseKeepAlive(keepAliveResponseStreamObserver);
+        requestObserver.onNext(newKeepAliveRequest(leaseId));
+        requestObserver.onCompleted();
+
+
+        return null;
+    }
+
+    /**
+     *Scan all the leases and seed keep alive request to etcd server
+     */
+    private void keepAliveExecutor(){
+        synchronized (lock){
+            long now = System.currentTimeMillis();
```

Comment:
Have you considered using nanos than millis? As the lease here is more about duration not absolute time, nanos will have better performance.